### PR TITLE
Xbutil2 bist

### PR DESCRIPTION
**Pull request by chienwei-lan**

### CODE REVIEW

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 1]
```diff
@@ -561,6 +562,69 @@ search_and_program_xclbin(const std::shared_ptr<xrt_core::device>& dev, boost::p
...
+{
+  int ret, bo_size = 0x1000;
+  xclBufferHandle boh = xclAllocBO(handle, bo_size, 0, XCL_BO_FLAGS_EXECBUF);
+
+  if (boh == NULLBO) {
+    _ptTest.put("status", "failed");
+    logger(_ptTest, "Error", "Couldn't allocate BO");
+    return false;
+  }
+  char *boptr = (char *)xclMapBO(handle, boh, true);
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:576]: please use c++ standard:
`auto boptr = reinterpret_cast<char*>(xclMapBO(handle, boh, true));`

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 2]
```diff
@@ -561,6 +562,69 @@ search_and_program_xclbin(const std::shared_ptr<xrt_core::device>& dev, boost::p
...
+
+  std::this_thread::sleep_for(std::chrono::seconds(sleep_secs));
+
+  if(!bist_alloc_execbuf_and_wait(handle, ERT_CLK_CALIB, _ptTest))
+    return false;
+
+  uint64_t end = xrt_core::device_query<xrt_core::query::clock_timestamp>(_dev);
+
+  /* Calculate the clock frequency in MHz*/
+  double freq = ((end + ULONG_MAX - start) & (ULONG_MAX)) / (sleep_secs*one_million);
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:622]: C++ standard:
`std::numeric_limits<unsigned long>::max()`

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 3]
```diff
@@ -561,6 +562,69 @@ search_and_program_xclbin(const std::shared_ptr<xrt_core::device>& dev, boost::p
...
+  }
+  while (ecmd->state < ERT_CMD_STATE_COMPLETED);
+
+  return true;
+}
+
+static bool 
+clock_calibration(const std::shared_ptr<xrt_core::device>& _dev, xclDeviceHandle handle, boost::property_tree::ptree& _ptTest)
+{
+  int sleep_secs = 2, one_million = 1000000;
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:607]: Are these static const?

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 4]
```diff
@@ -837,6 +901,25 @@ hostMemBandwidthKernelTest(const std::shared_ptr<xrt_core::device>& _dev, boost:
...
+ * TEST #10
+ */
+void
+bistTest(const std::shared_ptr<xrt_core::device>& _dev, boost::property_tree::ptree& _ptTest)
+{
+  if(!search_and_program_xclbin(_dev, _ptTest)) {
+    return;
+  }
+
+  std::string msg;
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:914]: is this used anywhere?

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 5]
```diff
@@ -837,6 +901,25 @@ hostMemBandwidthKernelTest(const std::shared_ptr<xrt_core::device>& _dev, boost:
...
+  xclbin_lock xclbin_lock(_dev);
+
+  if (clock_calibration(_dev, _dev->get_device_handle(), _ptTest))
+     _ptTest.put("status", "passed");
+
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:919]: remove extra line

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 6]
```diff
@@ -561,6 +562,69 @@ search_and_program_xclbin(const std::shared_ptr<xrt_core::device>& dev, boost::p
   return true;
 }
 
+static bool
+bist_alloc_execbuf_and_wait(xclDeviceHandle handle, enum ert_cmd_opcode opcode, boost::property_tree::ptree& _ptTest)
+{
+  int ret, bo_size = 0x1000;
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:568]: should bo size be static const?

**Review by AShivangi — APPROVED** [review]
Overall looks good! There are some minor suggestions for code improvement. 

Please hold off merging this until https://github.com/Xilinx/XRT/pull/4750 is merged. There might be some overlap and that PR is urgent.

**Comment by xilinxgitops** [comment]
retest this please

**Pull request merged**